inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp

```diff
@@ -654,12 +653,26 @@ void RemovePermutationsNHWCToNCHWPass::run() {
         auto next = getInputTo(l->outData.front()).begin()->second;
         auto prev = CNNNetPrevLayer(l);
 
-        if (!LayerInfo(next).isPermute() || !LayerInfo(prev).isPermute()) {
+        // The next layer must be NCHW to NHWC permute
+        if (!LayerInfo(next).isPermute() || next->GetParamAsInts("order") != std::vector<int>({0, 2, 3, 1})) {
             continue;
         }
 
-        if (getPassManager()->getPolicy().NHWCToNCHWPolicy == Policy::NHWCToNCHW::REMOVE_ALL) {
-            permutationsToRemove.push_back(prev);
+        // The previous layer must be NHWC to NCHW permute or have 1D data
+        if (LayerInfo(prev).isPermute()) {
+            if (prev->GetParamAsInts("order") != std::vector<int>({0, 3, 1, 2})) continue;
+            if (getPassManager()->getPolicy().NHWCToNCHWPolicy == Policy::NHWCToNCHW::REMOVE_ALL) {
+                permutationsToRemove.push_back(prev);
+            }
+        } else  {
+            if (prev->outData.size() != 1 || getInputTo(prev->outData[0]).size() != 1) {
+                continue;
+            }
+            int non_1_dims = 0;
+            for (auto dim : prev->outData[0]->getDims()) {
```

Comment:
I don't understand what is purpose of this loop.

Reply:
If only one dimension isn't equal to 1 in convolution input permute layer may not be inserted before convolution, so if we have permute after convolution and 1d input we should mark convolution layout as NHWC too and remove the permute after convolution.
I've replaced this loop with std::count_if.